### ISSUE TITLE
CellImg incorrect dimension exception

### ISSUE THREAD

**Issue by mstritt** [imglib/imglib2#267]
Hello,
when I try to read an image as a cell image using scifio I got a wrong dimensionality exception.
I tried it with ndpi files, but get the same error when opening simple jpg images.
Without the SCIFIOConfig.ImgMode.CELL config is works, but for larger images I need to read it in cell mode.

Example:
```  
import io.scif.config.SCIFIOConfig;
import io.scif.img.ImgOpener;
import io.scif.img.SCIFIOImgPlus;
import net.imglib2.type.numeric.real.FloatType;
import org.scijava.io.location.FileLocation;
import java.util.List;

public class TileReader2 {
    public static void main(String[] args) throws Exception{
        // NDPI file from http://downloads.openmicroscopy.org/images/Hamamatsu-NDPI/manuel/test3-DAPI%202%20(387)%20.ndpi
        String testPath = "test3-DAPI 2 (387) .ndpi";
        SCIFIOConfig config = new SCIFIOConfig();
        config.imgOpenerSetImgModes( SCIFIOConfig.ImgMode.CELL );   // without that line it works
        final List<SCIFIOImgPlus<FloatType>> img = new ImgOpener().openImgs(new FileLocation(testPath), new FloatType(), config );
        img.get(0).firstElement();   // fails
    }
}
```  

results in this exception:
```  
java.lang.IllegalArgumentException: Expected bounds of dimensionality 3 but was 2
	at io.scif.img.cell.loaders.AbstractArrayLoader.validateBounds(AbstractArrayLoader.java:306)
	at io.scif.img.cell.loaders.AbstractArrayLoader.read(AbstractArrayLoader.java:279)
	at io.scif.img.cell.loaders.AbstractArrayLoader.read(AbstractArrayLoader.java:270)
	at io.scif.img.cell.loaders.AbstractArrayLoader.read(AbstractArrayLoader.java:252)
	at io.scif.img.cell.loaders.AbstractArrayLoader.loadArray(AbstractArrayLoader.java:232)
	at io.scif.img.cell.SCIFIOCellImgFactory$SCIFIOCellLoader.load(SCIFIOCellImgFactory.java:200)
	at net.imglib2.cache.img.LoadedCellCacheLoader.get(LoadedCellCacheLoader.java:91)
	at net.imglib2.cache.img.LoadedCellCacheLoader.get(LoadedCellCacheLoader.java:51)
	at net.imglib2.cache.img.DiskCellCache.get(DiskCellCache.java:104)
	at net.imglib2.cache.img.DiskCellCache.get(DiskCellCache.java:43)
	at net.imglib2.cache.IoSync.get(IoSync.java:174)
	at net.imglib2.cache.ref.SoftRefLoaderRemoverCache.get(SoftRefLoaderRemoverCache.java:158)
```  

I use these gradle dependencies:

```   
    compile 'net.imglib2:imglib2:5.8.0'
    compile 'net.imglib2:imglib2-algorithm:0.11.1'
    compile 'net.imglib2:imglib2-algorithm-fft:0.2.0'
    compile 'net.imglib2:imglib2-algorithm-gpl:0.2.3'
    compile 'net.imglib2:imglib2-ij:2.0.0-beta6'
    compile 'net.imglib2:imglib2-realtransform:2.2.1'
    compile 'net.imglib2:imglib2-ui:2.0.0'
    compile 'net.imagej:imagej:2.0.0-rc-69'
    compile group: 'io.scif', name: 'scifio-bf-compat', version: '4.0.0'
    compile (group: 'ome', name:'formats-gpl', version:'5.9.2') {     // 6.1.1 5.9.2
        exclude (module: 'blitz')
    }
    compile 'ch.qos.logback:logback-classic:1.1.7'
    testCompile group: 'junit', name: 'junit', version: '4.12'
```  

I have already addressed this issue in the image.sc forum with further details:
[https://forum.image.sc/t/cellimg-incorrect-dimension-exception/27826](https://forum.image.sc/t/cellimg-incorrect-dimension-exception/27826)

Any ideas how to get this running?

Regards,
Manuel

**Comment by mstritt** [comment]
Hi @ctrueden,
thanks a lot for addressing this issue. I'm thankful to all kind of input here, and yes, with your hints I'll start another debugging session.

So far I figured out that the error comes from

```
CellCursor->private void moveToNextCell()
	{
		cursorOnCells.fwd();
		isNotLastCell = cursorOnCells.hasNext();
		lastIndexInCell = ( int ) ( getCell().size() - 1 );
		index = -1;
		type.updateContainer( this );
	}
```
-> getCell()

But I'll dig a bit deeper into it.

Background:
We'll rewrite mayor parts of the Orbit Image Analysis backend. The idea is to have a whole slide analysis framework which is based on imgLib2 - so that we can make use imgLib2.ops and reuse existing algorithms instead of rewriting things.

Cheers,
Manuel

**Comment by mstritt** [comment]
Hi @ctrueden , 
ok, so this is what I found so far:
in the `SCIFIOCellLoader` the load method calls `loader.loadArray(cell, data)` with cell being 
`SingleCellArrayImg [(0,0,0) -- (3967,87,0) = 3968x88x1]`
Then in the `AbstractArrayLoader` the read method calls
`validateBounds(getPlanarAxisLengths(reader.getMetadata()), bounds);`
with 
`getPlanarAxisLengths(reader.getMetadata()) = [3, 3968, 4864]`
and 
`bounds = FinalInterval [(0,0) -- (3967,87) = 3968x88]`

for this validateBounds fails because the `getPlanarAxisLengths()` has dimensionality 3 and bounds only 2.

It seems that `getPlanarAxisLengths()` sees the channel as one dimension, but bounds not.
Is it a problem with `getPlanarAxisLengths()`?
The image is a RGB image.

Thanks for any help here in advance!

Regards,
Manuel

**Comment by mstritt** [comment]
Hi @ctrueden  and @hinerm ,
After some further analysis the problem seems to be the planarAxisCount which is 3 instead of 2.
`reader.getMetadata().imageMeta.get(0).planarAxisCount=2`
but
`reader.getMetadata().meta.imageMeta.get(0).planarAxisCount=3`
so after calling `unrap()` in `AbstractArrayLoader`, `getPlanarAxisLengths()` returns 3 - but probably should be 2 ?

Regards,
Manuel

**Comment by axtimwalde** [comment]
Is this an ImgLib2 or a https://github.com/scifio/scifio issue?

**Comment by mstritt** [comment]
Good question. Probably it comes from the `SCIFIOCellLoader `which is in Scifio - should we create a new issue in scifio and reference this one here?

**Comment by axtimwalde** [comment]
Yes, I believe that makes more sense. scifio folks won't look for issues here.

**Comment by ctrueden** [comment]
Closed in favor of scifio/scifio#399.
